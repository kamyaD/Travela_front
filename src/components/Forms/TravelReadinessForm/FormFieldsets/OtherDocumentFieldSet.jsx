import React, { Component } from 'react';
import InputRenderer from '../../FormsAPI';
import formMetadata from '../../FormsMetadata/TravelReadinessMetaData/OtherDocumentMetaData';

class OtherDocumentFieldSet extends Component {
  render() {
    const {renderInput} = new InputRenderer(formMetadata);
    return (
      <fieldset>
        <div className="input-group visa-input">
          {renderInput('name', 'text')}
          {renderInput('dateOfIssue', 'date')}
          {renderInput(
            'documentId',
            'text',
            {
              labelNote: '(Optional)',
              required: false,
            }
          )}
          {renderInput('expiryDate', 'date')}
        </div>
      </fieldset>
    );
  }
}

export default OtherDocumentFieldSet;
