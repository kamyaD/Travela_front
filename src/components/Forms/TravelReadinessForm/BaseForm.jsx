import React, { Component } from 'react';
import {PropTypes} from 'prop-types';
import toast from 'toastr';
import axios from 'axios';
import { isEmpty } from 'lodash';
import moment from 'moment';
import {FormContext} from '../FormsAPI';
import './TravelDocument.scss';
import VisaFormFieldSet from './FormFieldsets/VisaFormFieldset';
import OtherDocumentFieldSet from './FormFieldsets/OtherDocumentFieldSet';
import FileUploadField from './FormFieldsets/FileUploadField';
import documentUpload from '../../../images/icons/document-upload-blue.svg';
import DocumentAPI from '../../../services/DocumentAPI';
import SubmitArea from './FormFieldsets/SubmitArea';
import Preloader from '../../Preloader/Preloader';

class BaseForm extends Component {
  constructor(props) {
    super(props);
    const { defaultFormState } = this.props;
    this.defaultState = defaultFormState;
    this.state = {...this.defaultState};
  }

  componentWillReceiveProps(nextProps, nextContext) {
    const { errors, document } = nextProps;
    const { document: currentDocument } = this.props;
    if ((isEmpty(currentDocument) && !isEmpty(document)) && document.type === 'visa') {
      const { country, entryType, visaType, dateOfIssue, expiryDate, cloudinaryUrl } = document.data;
      this.newState = {
        values: {
          country, entryType, visaType,
          dateOfIssue: moment(dateOfIssue),
          expiryDate: moment(expiryDate)
        },
        id: document.id,
        image: cloudinaryUrl,
      };
      this.setState({errors, uploadingDocument: false, ...this.newState });
    } else if ((isEmpty(currentDocument) && !isEmpty(document)) && document.type === 'other') {
      const { name, dateOfIssue, expiryDate, documentId, cloudinaryUrl } = document.data;
      this.newState = {
        values: {
          name,
          dateOfIssue: dateOfIssue === '' ? '' : moment(dateOfIssue),
          expiryDate: moment(expiryDate),
          documentId
        },
        id: document.id,
        image: cloudinaryUrl,
      };
      this.setState({errors, uploadingDocument: false, ...this.newState });
    }
    this.setState({errors, uploadingDocument: false});
  }

  componentWillUnmount() {
    const { fetchUserData, user } = this.props;
    fetchUserData(user.currentUser.userId);
  }

  onCancel = (event) => {
    event.preventDefault();
    const { closeModal } = this.props;
    closeModal();
  };

  handleSubmit = async (event) => {
    event.preventDefault();
    const {  values, image, errors, documentUploaded, id } = this.state;
    const { createTravelReadinessDocument, editTravelReadinessDocument, modalType, documentType } = this.props;
    const { dateOfIssue, expiryDate } = values;
    if(documentUploaded){

      const newValues = {
        ...values,
        dateOfIssue: dateOfIssue === '' ? '' : moment(dateOfIssue).format('YYYY/MM/DD'),
        expiryDate: moment(expiryDate).format('YYYY/MM/DD')
      };

      if (modalType === `edit ${documentType}`) {
        editTravelReadinessDocument(documentType, newValues, id);
      } else {
        createTravelReadinessDocument(documentType, newValues);
      }
    }else {
      if (image) {
        const fd = new FormData();
        fd.append('file', image);
        fd.append('upload_preset', process.env.REACT_APP_PRESET_NAME);

        try {
          delete axios.defaults.headers.common['Authorization'];
          this.setState({ uploadingDocument: true });
          const imageData = await axios.post(process.env.REACT_APP_CLOUNDINARY_API, fd);
          const {data: {url}} = imageData;
          this.setState({
            documentUploaded: true,
            uploadingDocument: false,
            values: {...values, cloudinaryUrl: url}
          });
          DocumentAPI.setToken();
          const documentValues = {
            ...values, 
            cloudinaryUrl: url,
            dateOfIssue: dateOfIssue === '' ? '' : moment(dateOfIssue).format('MM/DD/YYYY'),
            expiryDate: moment(expiryDate).format('MM/DD/YYYY')
          };

          if (modalType === `edit ${documentType}`) {
            editTravelReadinessDocument(documentType, { ...documentValues }, id);
          } else {
            createTravelReadinessDocument(documentType, { ...documentValues });
          }
        } catch (e) {
          toast.error('Error uploading passport. Please try again!');
        }
      } else {
        this.setState({errors: {...errors, cloudinaryUrl: 'Please upload a passport'}});
      }
    }
  };


  handleUpload = (e) => {
    e.preventDefault();
    const image = e.target.files[0];
    if (!['image/jpeg', 'image/png'].includes(image.type)) {
      return toast.error('Invalid file type. Please upload an image');
    }
    if (image.size>Math.pow(10,7)) {
      return toast.error('File is too large');
    }
    const {  name } = image;
    this.setState({name, image, imageChanged: true});
  };

  render() {
    const {errors, values, hasBlankFields, uploadingDocument, name, imageChanged } = this.state;
    const { documentType, modalType, document, fetchingDocument } = this.props;
    if (documentType === 'other') delete errors.documentid;
    return(
      <div>
        {fetchingDocument ? <Preloader /> : (
          <FormContext values={values} errors={errors} targetForm={this}>
            <form className="travel-document-form" onSubmit={this.handleSubmit}>
              { documentType === 'visa' &&  <VisaFormFieldSet /> }
              { documentType === 'other' &&  <OtherDocumentFieldSet /> }
              <div className="travel-document-select-file">
                <p>Attach File</p>
                <FileUploadField 
                  name={name}
                  documentUpload={documentUpload} 
                  handleUpload={this.handleUpload}
                  document={document}
                  modalType={modalType}
                />
              </div>
              <hr />
              <div className="travel-document-submit-area">
                <SubmitArea
                  onCancel={this.onCancel} hasBlankFields={hasBlankFields && !imageChanged} 
                  send={
                    (modalType === 'edit other' || modalType === 'edit visa') ? 'Save Changes' : 
                      (modalType === 'add other' ? 'Add Document' : 'Add Visa')}
                  loading={uploadingDocument} />
              </div>
            </form>
          </FormContext>)}
      </div>
    );
  }
}

BaseForm.propTypes = {
  errors: PropTypes.object,
  createTravelReadinessDocument: PropTypes.func.isRequired,
  closeModal: PropTypes.func.isRequired,
  fetchUserData: PropTypes.func.isRequired,
  user: PropTypes.object.isRequired,
  documentType: PropTypes.string.isRequired,
  defaultFormState: PropTypes.object,
  editTravelReadinessDocument: PropTypes.func,
  document: PropTypes.object,
  modalType: PropTypes.string, fetchingDocument: PropTypes.bool
};

BaseForm.defaultProps = {
  editTravelReadinessDocument: () => {},
  document: {},
  modalType: '',
  errors: {},
  defaultFormState: {},
  fetchingDocument: false
};
export default BaseForm;
