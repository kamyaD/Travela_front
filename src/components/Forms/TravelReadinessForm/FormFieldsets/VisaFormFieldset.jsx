import React, { Component } from 'react';
import moment from 'moment';
import {PropTypes} from 'prop-types';
import '../TravelDocument.scss';
import { uniq } from 'lodash';
import countries  from 'world-countries';
import InputRenderer from '../../FormsAPI';
import formMetadata from '../../FormsMetadata/TravelReadinessMetaData/NewVisaMetadata';
import { reasonsWarningColor } from '../../NewRequestForm/FormFieldsets/TravelDetailsItem';

class VisaFormFieldSet extends Component {
  renderCharcters = (reasonsLimit, errorMessagePosition, visaType) => {
    return (
      <div
        className={`character__conditions__visa ${errorMessagePosition}`}
        style={{ color: `${reasonsLimit.color}`, display: `${visaType === 'Other' ? 'block' : 'none'}`}}
      >
        {reasonsLimit.charLeft}
      </div>
    );
  }
  getCountry = () => {
    const countryChoices = countries.map(country => country.name.common);
    return uniq(countryChoices);
  }

  render() {
    const {renderInput} = new InputRenderer(formMetadata);
    const { visaType, onChangeVisa, otherVisaTypeDetails, values: { country } } = this.props;
    const characters = otherVisaTypeDetails;
    let charLength = characters ? characters.trim().length : '';
    let reasonsLimit = reasonsWarningColor(charLength, 140);
    let errorMessagePosition = charLength > 70 ? 'error__height-two' : 'error__height-one';
    return (
      <fieldset>
        <div className="input-group visa-input">
          {renderInput('country', 'filter-dropdown-select',
            { choices: this.getCountry(), size: '', value: country }
          )}
          {renderInput('entryType', 'dropdown-select', { size: '' })}
          {renderInput('dateOfIssue', 'date', {
            maxDate: moment(), showYearDropdown: true
          })}
          {renderInput('expiryDate', 'date', {
            minDate: moment(),
            showYearDropdown: true
          })}
          {renderInput('visaType', 'dropdown-select', {
            size: '',
            onChange: onChangeVisa,
            value: visaType
          })}
          {visaType == 'Other'  && renderInput('otherVisaTypeDetails', 'textarea', {
            maxLength:'140',
            rows: '20',
            className: 'other-visa-description'
          })}
          {this.renderCharcters(reasonsLimit, errorMessagePosition, visaType)}
        </div>
      </fieldset>
    );
  }
}

VisaFormFieldSet.propTypes={
  visaType: PropTypes.string,
  onChangeVisa: PropTypes.func.isRequired,
  otherVisaTypeDetails: PropTypes.string,
  values: PropTypes.object
};

VisaFormFieldSet.defaultProps = {
  visaType: '',
  otherVisaTypeDetails: '',
  values: {}
};

export default VisaFormFieldSet;
