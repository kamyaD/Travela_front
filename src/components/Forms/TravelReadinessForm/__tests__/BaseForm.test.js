import React from 'react';
import toast from 'toastr';
import moxios from 'moxios';
import BaseForm from '../BaseForm';
import VisaFormFieldset from '../FormFieldsets/VisaFormFieldset';
import OtherDocumentFieldSet from '../FormFieldsets/OtherDocumentFieldSet';



const props = {
  errors: {},
  createTravelReadinessDocument: jest.fn(),
  closeModal: jest.fn(),
  fetchUserData: jest.fn(),
  user: {}
};

const otherDocumentDefaultFormState  = {
  values: {
    name: '',
    dateOfIssue: '',
    expiryDate: '',
  },
  errors: {},
  hasBlankFields: true,
  isSubmitting: false,
  uploadProgress: 0,
  cloudinaryUrl: ''
};

const visaDefaultFormState  = {
  values: {
    name: '',
    dateOfIssue: '',
    expiryDate: '',
  },
  value: {
    name: '',
    dateOfIssue: '',
    expiryDate: '',
  },
  errors: {},
  hasBlankFields: true,
  isSubmitting: false,
  uploadProgress: 0,
  cloudinaryUrl: ''
};

const textFile = new Blob(['This is a text file'], {type : 'text/plain'});
textFile.name = 'textFile.txt';

const validFile = new Blob(['This is a valid png file'], {type : 'image/png', size: 1092});
validFile.name = 'file.png';

toast.error = jest.fn();
const event = {
  preventDefault: jest.fn(),
  target:{
    files: [validFile]
  }
};

describe('<OtherDocumentForm />', () => {
  const wrapper = mount(
    <BaseForm {...props} documentType="other" defaultFormState={otherDocumentDefaultFormState} />
  );
  process.env.REACT_APP_CLOUNDINARY_API = 'https://api.cloudinary.com/v1_1/skybound/image/upload';
  beforeEach(() => {
    moxios.install();
  });

  afterEach(() => {
    moxios.uninstall();
    toast.error.mockReset();
  });
  it('renders without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders other document modal', () => {
    const wrapperWithOtherDocumentField = mount(
      <BaseForm
        {...props}
        documentType="other"
        defaultFormState={otherDocumentDefaultFormState}
      />
    );
    expect(
      wrapperWithOtherDocumentField.find(OtherDocumentFieldSet))
      .toHaveLength(1);
  });

  it('renders other document modal with button text \'Add Document\'', () => {
    const wrapperWithOtherDocumentField = mount(
      <BaseForm
        {...props}
        documentType="other"
        defaultFormState={otherDocumentDefaultFormState}
      />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Add Document');
  });

  it('renders visa  modal', () => {
    const wrapperWithOtherDocumentField = mount(
      <BaseForm {...props} documentType="visa" defaultFormState={visaDefaultFormState} />
    );
    expect(
      wrapperWithOtherDocumentField.find(VisaFormFieldset))
      .toHaveLength(1);
  });

  it('toasts an error if cloudinary returns an error', () => {
    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, { status: 500});
    wrapper.find('#select-file').simulate('change', event);
  });

  it('renders visa  modal with button text \'Add Visa\'', () => {
    const wrapperWithOtherDocumentField = mount(
      <BaseForm {...props} documentType="visa" defaultFormState={visaDefaultFormState} />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Add Visa');
  });

  it('renders visa  default value \'Add\' on the button', () => {
    const wrapperWithOtherDocumentField = mount(
      <BaseForm {...props} documentType="none" defaultFormState={visaDefaultFormState} />
    );
    expect(
      wrapperWithOtherDocumentField.find('button#submit').text()).toEqual('Add ');
  });

  it('should display an error when no value is input', () => {
    const formWrapper = mount(
      <BaseForm {...props} documentType="visa" defaultFormState={visaDefaultFormState} />
    );
    let isSubmitted = false;
    formWrapper.find('form').simulate('submit', {
      preventDefault: () => {
        isSubmitted = true;
      },
    });
    expect(isSubmitted).toBe(true);
  });

  it('should toast an error if the document file does not upload', () => {
    wrapper.setState({values: {...document}});
    wrapper.find('#select-file').simulate('change', event);

    process.env.REACT_APP_CLOUNDINARY_API = 'https://mock-passport-cloudinary-api-succeed';

    const cloudinaryResponse = {
      url: 'secure url'
    };

    moxios.stubRequest(process.env.REACT_APP_CLOUNDINARY_API, {
      status: 404,
      response: cloudinaryResponse
    });

    wrapper.setState({documentUploaded: false});
    wrapper.find('.travel-document-form').simulate('submit');
    expect(toast.error).toHaveBeenCalledTimes(0);
  });

  it('should display errors when received from the api', () => {
    wrapper.setProps({ errors: { cloudinaryUrl: 'Please provide a document'}});
    expect(wrapper.state().errors).toEqual({cloudinaryUrl: 'Please provide a document'});
  });
});
