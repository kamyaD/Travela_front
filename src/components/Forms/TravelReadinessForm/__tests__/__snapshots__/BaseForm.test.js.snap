// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OtherDocumentForm /> renders without crashing 1`] = `
<form
  className="travel-document-form"
  onSubmit={[Function]}
>
  <fieldset>
    <div
      className="input-group visa-input"
    >
      <div
        className="form-input  "
      >
        <label
          htmlFor="name"
        >
          Name
          <span
            style={
              Object {
                "color": "red",
              }
            }
          >
            *
          </span>
        </label>
        <input
          autoComplete="off"
          className=" "
          label="Name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
          value=""
        />
        <span
          className="error"
        />
      </div>
      <div
        className="form-input  "
      >
        <label
          htmlFor="dateOfIssue"
        >
          Date of Issue
          <span
            style={
              Object {
                "color": "red",
              }
            }
          >
            *
          </span>
        </label>
        <div
          className="date-wrapper "
          id="dateOfIssue_date"
        >
          <div>
            <div
              className="react-datepicker-wrapper"
            >
              <div
                className="react-datepicker__input-container"
              >
                <input
                  autoComplete="off"
                  className=""
                  disabled={false}
                  name="dateOfIssue"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <img
            alt="cal"
            className="calendar-icon"
            src="calendar_icon.svg"
          />
        </div>
        <span
          className="error"
        />
      </div>
      <div
        className="form-input  "
      >
        <label
          htmlFor="documentId"
        >
          Document Id
          <span>
            Â 
            (Optional)
          </span>
        </label>
        <input
          autoComplete="off"
          className=" "
          label="Document Id"
          name="documentId"
          onBlur={[Function]}
          onChange={[Function]}
          required={false}
          type="text"
        />
        <span
          className="error"
        />
      </div>
      <div
        className="form-input  "
      >
        <label
          htmlFor="expiryDate"
        >
          Expiry Date
          <span
            style={
              Object {
                "color": "red",
              }
            }
          >
            *
          </span>
        </label>
        <div
          className="date-wrapper "
          id="expiryDate_date"
        >
          <div>
            <div
              className="react-datepicker-wrapper"
            >
              <div
                className="react-datepicker__input-container"
              >
                <input
                  autoComplete="off"
                  className=""
                  disabled={false}
                  name="expiryDate"
                  onBlur={[Function]}
                  onChange={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onKeyDown={[Function]}
                  readOnly={false}
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
          <img
            alt="cal"
            className="calendar-icon"
            src="calendar_icon.svg"
          />
        </div>
        <span
          className="error"
        />
      </div>
    </div>
  </fieldset>
  <div
    className="travel-document-select-file"
  >
    <p>
      Attach File
    </p>
    <div
      className="document-input__input-container"
    >
      <label
        className="document-input__input-container__prompts"
        htmlFor="select-file"
      >
        <img
          alt="Document Upload"
          className="document-input__input-container__prompts__img"
          src="document-upload-blue.svg"
        />
        <div
          className="document-input__input-container__prompts__text"
        >
          <p>
            Drag file here or choose from computer
          </p>
        </div>
        <input
          id="select-file"
          onChange={[Function]}
          type="file"
        />
      </label>
    </div>
  </div>
  <hr />
  <div
    className="travel-document-submit-area"
  >
    <fieldset>
      <div
        className="submit-area"
      >
        <button
          className="bg-btn bg-btn--inactive"
          id="cancel"
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
        <button
          className="bg-btn bg-btn--active"
          disabled={true}
          id="submit"
          type="submit"
        >
          Add Visa
        </button>
      </div>
    </fieldset>
  </div>
</form>
`;
